knowledge_repo/app/models.py

```diff
@@ -234,26 +235,59 @@ class Vote(db.Model):
 
 
 @unique_constructor(
-    lambda username: username,
-    lambda query, username: query.filter(User.username == username)
+    lambda identifier: identifier,
+    lambda query, identifier: query.filter(User.identifier == identifier)
 )
-class User(db.Model):
+class User(db.Model, UserMixin):
     __tablename__ = 'users'
 
     id = db.Column(db.Integer, primary_key=True)
-    username = db.Column(db.String(500))
     created_at = db.Column(db.DateTime, default=func.now())
 
+    identifier = db.Column(db.String(500))  # Unique identifier across all login methods
+
+    username = db.Column(db.String(500))  # Username used to log in (may differ from identifier)
+    password = db.Column(db.String(500))  # Password for local logins
+
+    name = db.Column(db.String(500))  # Name as determined by auth method
```

Comment:
**identifier** is the username/email/identifier used to indicate the authors of posts in the YAML headers. Exactly what it is (email, LDAP username, some combination) is left up to the server administrator to decide. My recommendation for a default would be the author's email, since that means that posts can be handed around between different knowledge repo instances and have very little chance of clashing with other users; however, I was hesitant to force that on people. Internally we've been using LDAP usernames as the identifier.

Do you think we should force identifiers to be emails? That could be good for the knowledge post ecosystem in the long term.

**username** is the username provided to the Auth provider, which may be different than the **identifier**, especially if deployers set up multiple authentication methods. (i.e. You can log in with Google OR with LDAP.) It is mainly here as a way to allow administrators to have a *local* login method that does not require external logins, but where the username does not match the identifier.

**name** is the literal name of the person. I would be "Matthew Wardrop", or "M Wardrop" or "Matthew"; depending on how the authentication method chooses to set this field. The **preferred_name** is a preference field for users to override this default (not currently exposed in UI, since this PR was big enough), which is especially useful in places where the authentication provider cannot properly render names (such as CJK names or unusual non-Latin names).